src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php

```diff
@@ -81,24 +73,80 @@ public function finishView(FormView $view, FormInterface $form, array $options)
      */
     public function configureOptions(OptionsResolver $resolver)
     {
-        $entryOptionsNormalizer = function (Options $options, $value) {
-            $value['block_name'] = 'entry';
+        $entryOptionsNormalizer = function (Options $options, $entryOptions) {
+            if (is_callable($entryOptions)) {
+                // Wrap the callable to override 'block_name' option value
+                return function ($entry) use ($entryOptions) {
+                    $dynamicOptions = call_user_func($entryOptions, $entry);
+
+                    if (!is_array($dynamicOptions)) {
+                        throw new InvalidOptionsException(sprintf('The "entry_options" callable should return an array but got %s', is_object($dynamicOptions) ? get_class($dynamicOptions) : gettype($dynamicOptions)));
+                    }
+
+                    $dynamicOptions['block_name'] = 'entry';
+
+                    return $dynamicOptions;
+                };
+            }
+
+            // Else $entryOption is an array
+            $entryOptions['block_name'] = 'entry';
+
+            return $entryOptions;
+        };
+
+        $prototypeOptionsNormalizer = function (Options $options, $value) {
+            if (is_callable($options['entry_options'])) {
+                // Pass the prototype data to the callable
+                $entryOptions = call_user_func(
+                    $options['entry_options'],
+                    // BC, 'prototype_data' will be removed in 4.0.
+                    $options['prototype_data'] ?: (isset($value['data']) ? $value['data'] : null)
+                );
+            } else {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method configureOptions uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://www.codacy.com/app/git_9/symfony/file/2332080378/issues/source?bid=3207821&fileBranchId=3207880#l106)